### ISSUE TITLE
User-generated exceptions thrown from within RHS not surfacing in CLJ

### ISSUE THREAD

**Issue by alex-dixon** [cerner/clara-rules#320]
Precept's `insert!` wrapper now throws an error when the user attempts to insert contradicting facts logically. I have it working in CLJS, but in Clojure I think Clara catches it and producing a different error in response. 

So far I haven't been able to figure out how to throw the error in such a way that Clara does not catch the Exception. I believe the relevant line is https://github.com/cerner/clara-rules/blob/master/src/main/clojure/clara/rules/engine.cljc#L1771. 

Any help on how I might be able to work around this would be appreciated. :)

**Comment by WilliamParker** [comment]
The intent here is to wrap exceptions thrown in a rule RHS in an ExceptionInfo exception that contains information about the environment in which they were thrown (which rule, what facts caused the rule to fire, etc.)  The original exception is [added as the cause on the ExceptionInfo](https://github.com/cerner/clara-rules/blob/0.15.0/src/main/clojure/clara/rules/engine.cljc#L1795) though so it should be accessible; see the [three-argument doc for ex-info](https://clojuredocs.org/clojure.core/ex-info).  That cause is then accessible via the [getCause method](https://docs.oracle.com/javase/8/docs/api/java/lang/Throwable.html#getCause--).

If I'm misunderstanding, could you give a concrete example of the problem you're running into?

**Comment by alex-dixon** [comment]
precept.util
```clj

**Comment by alex-dixon** [comment]
Sorry...appears to be working now. I'm really not sure why. Let me know if you would like more details. Sorry for any time I wasted.

**Comment by mrrodriguez** [comment]
@alex-dixon I may be misreading or misunderstanding a part of what you are saying, but it sounds like Clara is still throwing an exception, it's just an exception that "wraps" your exception.

If that is the case, I'd say that is fairly typical behavior as far as how exceptions go. Any time you are throwing an exception from some other lib's code/context, I don't think it should be assumed that your exception will "bubble out" without any decoration/wrapping.

Typically you catch and rethrow exceptions with more context-specific info, which is what Clara is doing with RHS exceptions. Again, perhaps I'm just completely misunderstanding the issue though.

I think Clara should (and I believe it is) keep a reference to your exception on the cause-stack though. You could write or use something that walks the stack to find the exception you are looking for though. It typically just means walking the cause-chain. I'm pretty sure you can do that in clj (you can) and cljs (I think).

**Comment by alex-dixon** [comment]
Ok. I don't think either of you are misunderstanding at all. I'm new to writing libraries and Java also. If there's a way to surface the error I'm trying to throw then there's no issue here, just wanted to make sure it was possible before going down that path. I'll try figuring out a solution that incorporates what you suggest. Thanks!

**Comment by WilliamParker** [comment]
@alex-dixon no problem.  Just another note - much if not most of the JVM-world tooling I've seen will print the exception chain in one form or another, not just its "end".  So if your functional need is just for your specific exception to be printed when an error is thrown you may not actually need to do any specific handling.  This is tool-dependent though.  By the chain I mean something like E has E1 as its cause and E1 has E2 as its cause, so tooling could print something like

````
E 
stack trace
caused by E1
stack trace
caused by E2
stack trace
````
